### PR TITLE
shader/execution: New access and constructor tests

### PR DESCRIPTION
**Pull request by ben-clayton**
* Fix `fromStorage()` for `vecN<bool>` inputs - this was missing parentheses, and without the parentheses complex expressions can incorrectly associate with the RHS, instead of the intended whole-expression.
* Add Add `bool` array and `vecN<bool>` access tests.
* Add runtime-sized array access tests.
* Add matrix access tests.
* Add zero-init constructor tests
* Add value constructor tests

Issue: #1606
Issue: #3170
Issue: #3171
Issue: #3172

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)
- [x] Test have be tested with compatibility mode validation enabled and behave as expected. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+g.test('vector_mix')
+  .specURL('https://www.w3.org/TR/WGSL/#value-constructor-builtin-function')
+  .desc(
+    `Test that a vector constructed from a mix of element values and sub-vectors produces the expected value`
+  )
+  .params(u =>
+    u
+      .combine('inputSource', allInputSources)
+      .combine('type', ['bool', 'i32', 'u32', 'f32', 'f16'] as const)
+      .combine('signature', [
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts:196]: This is nice and concise!

File: src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+    const nextValue = () => {
+      const value = elementValue(width++);
+      elements.push(value);
+      return elementType.create(value);
+    };
+    const args: Value[] = [];
+    for (const c of t.params.signature) {
+      switch (c) {
+        case '2':
+          args.push(vec2(nextValue(), nextValue()));
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts:227]: relies on parameter evaluation order.
 ECMAscript specifies left-to-right.

LGTM

File: src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+      arrayType,
+      t.params,
+      [
+        {
+          input: elements.map(e => elementType.create(e)),
+          expected: arrayType.create(elements),
+        },
+      ]
+    );
+  });
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts:419]: Please add .unimplemented() tests for structures.

File: src/webgpu/shader/execution/expression/constructor/zero_value.spec.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    const type = Type.array(t.params.length, Type[t.params.type]);
+    await run(
+      t,
+      basicExpressionBuilder(ops => `${type}()`),
+      [],
+      type,
+      { inputSource: 'const' },
+      [{ input: [], expected: type.create(0) }]
+    );
+  });
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/constructor/zero_value.spec.ts:107]: Please add .unimplemented() for structure variants

**Review by dneto0 — CHANGES_REQUESTED** [review]
Good overall.

Please add `.unimplemented()` tests for structure construction and zero-value.

File: src/webgpu/util/compare.ts [hunk 5]
```diff
@@ -118,7 +119,7 @@ function compareValue(got: Value, expected: Value): Comparison {
     };
   }
 
-  if (got instanceof VectorValue) {
+  if (got instanceof VectorValue || got instanceof ArrayValue) {
     const e = expected as VectorValue;
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply src/webgpu/util/compare.ts]: Good spot. Done.

File: src/webgpu/shader/execution/expression/expression.ts [hunk 6]
```diff
@@ -182,11 +182,22 @@ function storageType(ty: Type): Type {
...
+    return Type.array(ty.count, storageType(ty.elementType));
+  }
   return ty;
 }
 
+/** Structure used to hold [from|to]Storage conversion helpers  */
+type TypeConversionHelpers = {
+  // The module-scope WGSL to emit with the shader.
+  wgsl: string;
+  // A function that generates a unique identifier.
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply src/webgpu/shader/execution/expression/expression.ts:195]: Done

File: src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts (hunk 3 above)
Comment (ben-clayton) [reply src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts:419]: Done

File: src/webgpu/shader/execution/expression/constructor/zero_value.spec.ts (hunk 4 above)
Comment (ben-clayton) [reply src/webgpu/shader/execution/expression/constructor/zero_value.spec.ts:107]: Done

**Comment by ben-clayton** [comment]
Friendly ping!

**Comment by zoddicus** [comment]
Continues to lgtm, @dneto0 can you do an approval pass?

File: src/webgpu/shader/execution/expression/expression.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (dneto0) [reply src/webgpu/shader/execution/expression/expression.ts:195]: Not done?

**Review by dneto0 — APPROVED** [review]
Nit:  Missed one bit in a comment, but it's low stakes.

**Pull request merged**